File: Prims and Krushkals Algos/dijkstra.c
#include<stdio.h>
// #include <iostream>
#include <limits.h>
#define V 5
int findKey(int visited[], int distance[]) {
    int min = INT_MAX;
    int key = -1;

    for(int i = 0; i < V; i++) {
        if(visited[i] == 0 && distance[i] < min){
            min = distance[i];
            key = i;
        }
    }

    return key;

}

void dijkstra(int graph[V][V]){
    int visited[V] = {0};
    int distance[V];
    for (int i = 0; i < V; i++) distance[i] = INT_MAX;



    distance[0] = 0;
    for(int i = 0; i < V - 1; i++) {
        int u = findKey(visited, distance);
        visited[u] = 1;
       for(int j = 0; j < V; j++) {
            if(graph[u][j] != 0  && !visited[j] && distance[j] > distance[u] + graph[u][j]){
                distance[j] = distance[u] + graph[u][j];
            }
       } 
    }
    printf("\nShortest distances from node 0:\n");
    for (int i = 0; i < V; i++) {
        printf("Node %d: %d\n", i, distance[i]);
    }
}

int main() {
    int graph[V][V] = {
        // {0, 9, 75, 0, 0},
        // {0, 0, 95, 0, 42},
        // {0, 0, 0, 51, 0},
        // {0, 19, 0, 0},
        // {0, 0, 0, 0, 31}
        {0, 9, 75, 0, 0},
        {0, 0, 95, 0, 42},
        {0, 0, 0, 51, 0},
        {0, 19, 0, 0},
        {0, 0, 0, 0, 31}

    };

    printf("Graph:\n");
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            printf("%d\t", graph[i][j]);
        }
        printf("\n");
    }

    dijkstra(graph);

    return 4;
}




// #include <stdio.h>
// #include <limits.h>
// #define V 5

// int findKey(int visited[], int distance[]) {
//     int min = INT_MAX;
//     int key = -1;
//     for (int i = 0; i < V; i++) {
//         if (visited[i] == 0 && distance[i] < min) {
//             min = distance[i];
//             key = i;
//         }
//     }
//     return key;
// }

// void dijkstra(int graph[V][V]) {
//     int visited[V] = {0};
//     int distance[V];
//     for (int i = 0; i < V; i++) distance[i] = INT_MAX;

//     distance[0] = 0; // Start node distance = 0

//     for (int i = 0; i < V - 1; i++) {
//         int u = findKey(visited, distance);
//         visited[u] = 1;

//         for (int j = 0; j < V; j++) {
//             if (graph[u][j] != 0 && !visited[j] &&
//                 distance[u] + graph[u][j] < distance[j]) {
//                 distance[j] = distance[u] + graph[u][j];
//             }
//         }
//     }

//     printf("\nShortest distances from node 0:\n");
//     for (int i = 0; i < V; i++) {
//         printf("Node %d: %d\n", i, distance[i]);
//     }
// }

// int main() {
//     int graph[V][V] = {
//         {0, 9, 75, 0, 0},
//         {0, 0, 95, 0, 42},
//         {0, 0, 0, 51, 0},
//         {0, 19, 0, 0, 0},
//         {0, 0, 0, 0, 31}
//     };

//     printf("Graph:\n");
//     for (int i = 0; i < V; i++) {
//         for (int j = 0; j < V; j++) {
//             printf("%d\t", graph[i][j]);
//         }
//         printf("\n");
//     }

//     dijkstra(graph);

//     return 0;
// }

File: Dijkstra algo/Dijkstra.java
//static

import java.util.*;
public class Dijkstra{
    static final int V = 5; 
    static final int INF = Integer.MAX_VALUE;


    static int findKey(boolean[] visited, int[] distance) {
        int min = INF;
        int key = -1;

        for (int i = 0; i < V; i++) {
            if (!visited[i] && distance[i] < min) {
                min = distance[i];
                key = i;
            }
        }
        return key;
    }

    // Dijkstra’s algorithm
    static void dijkstra(int[][] graph) {
        boolean[] visited = new boolean[V];
        int[] distance = new int[V];
        Arrays.fill(distance, INF);

        distance[0] = 0; 

        for (int i = 0; i < V - 1; i++) {
            int u = findKey(visited, distance);
            if (u == -1) break; 

            visited[u] = true;

            for (int v = 0; v < V; v++) {
                if (graph[u][v] != 0 && !visited[v] && distance[u] != INF
                        && distance[v] > distance[u] + graph[u][v]) {
                    distance[v] = distance[u] + graph[u][v];
                }
            }
        }

        System.out.println("\nShortest distances from node 0:");
        for (int i = 0; i < V; i++) {
            if (distance[i] == INF)
                System.out.println("Node " + i + ": INF");
            else
                System.out.println("Node " + i + ": " + distance[i]);
        }
    }

    public static void main(String[] args) {
        int[][] graph = {
            {0, 9, 75, 0, 0},
            {0, 0, 95, 0, 42},
            {0, 0, 0, 51, 0},
            {0, 19, 0, 0, 0},
            {0, 0, 0, 0, 31}
        };

        System.out.println("Graph:");
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                System.out.print(graph[i][j] + "\t");
            }
            System.out.println();
        }

        dijkstra(graph);
    }
}


//dynamic code :

import java.util.*;

public class Dijkstra {

    static final int INF = Integer.MAX_VALUE;

    static int findKey(boolean[] visited, int[] distance, int V) {
        int min = INF;
        int key = -1;

        for (int i = 0; i < V; i++) {
            if (!visited[i] && distance[i] < min) {
                min = distance[i];
                key = i;
            }
        }
        return key;
    }

    static void dijkstra(int[][] graph, int src) {
        int V = graph.length;  
        boolean[] visited = new boolean[V];
        int[] distance = new int[V];
        Arrays.fill(distance, INF);

        distance[src] = 0; 

        for (int i = 0; i < V - 1; i++) {
            int u = findKey(visited, distance, V);
            if (u == -1) break; 

            visited[u] = true;

            for (int v = 0; v < V; v++) {
                if (graph[u][v] != 0 && !visited[v] && distance[u] != INF
                        && distance[v] > distance[u] + graph[u][v]) {
                    distance[v] = distance[u] + graph[u][v];
                }
            }
        }

        System.out.println("\nShortest distances from node " + src + ":");
        for (int i = 0; i < V; i++) {
            if (distance[i] == INF)
                System.out.println("Node " + i + ": INF");
            else
                System.out.println("Node " + i + ": " + distance[i]);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int V = sc.nextInt();

        int[][] graph = new int[V][V];
        System.out.println("Enter adjacency matrix (0 if no edge):");
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                graph[i][j] = sc.nextInt();
            }
        }

        System.out.print("Enter source node (0 to " + (V - 1) + "): ");
        int src = sc.nextInt();

        System.out.println("\nGraph:");
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                System.out.print(graph[i][j] + "\t");
            }
            System.out.println();
        }

        dijkstra(graph, src);

        sc.close();
    }
}

